File: crates/breez-sdk/bindings/examples/cli/langs/golang/contacts.go
package main

import (
	"flag"
	"fmt"
	"sort"

	breez_sdk_spark "github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
	"github.com/chzyer/readline"
)

// ContactCmd represents a single contacts subcommand.
type ContactCmd struct {
	Name        string
	Description string
	Run         func(sdk *breez_sdk_spark.BreezSdk, rl *readline.Instance, args []string) error
}

// ContactCommandNames lists all contacts subcommand names (used for REPL completion).
var ContactCommandNames = []string{
	"contacts add",
	"contacts update",
	"contacts delete",
	"contacts list",
}

// BuildContactRegistry returns a map of contacts subcommand name → ContactCmd.
func BuildContactRegistry() map[string]ContactCmd {
	return map[string]ContactCmd{
		"add": {
			Name:        "add",
			Description: "Add a new contact",
			Run:         handleContactAdd,
		},
		"update": {
			Name:        "update",
			Description: "Update an existing contact",
			Run:         handleContactUpdate,
		},
		"delete": {
			Name:        "delete",
			Description: "Delete a contact",
			Run:         handleContactDelete,
		},
		"list": {
			Name:        "list",
			Description: "List contacts",
			Run:         handleContactList,
		},
	}
}

// DispatchContactCommand dispatches a contacts subcommand.
func DispatchContactCommand(args []string, sdk *breez_sdk_spark.BreezSdk, rl *readline.Instance) {
	registry := BuildContactRegistry()

	if len(args) == 0 || args[0] == "help" {
		fmt.Println("\nContacts subcommands:")
		names := make([]string, 0, len(registry))
		for name := range registry {
			names = append(names, name)
		}
		sort.Strings(names)
		for _, name := range names {
			cmd := registry[name]
			fmt.Printf("  contacts %-30s %s\n", name, cmd.Description)
		}
		fmt.Println()
		return
	}

	subName := args[0]
	subArgs := args[1:]

	cmd, ok := registry[subName]
	if !ok {
		fmt.Printf("Unknown contacts subcommand: %s. Use 'contacts help' for available commands.\n", subName)
		return
	}

	if err := cmd.Run(sdk, rl, subArgs); err != nil {
		fmt.Printf("Error: %v\n", err)
	}
}

// ---------------------------------------------------------------------------
// Contacts command handlers
// ---------------------------------------------------------------------------

// --- add ---

func handleContactAdd(sdk *breez_sdk_spark.BreezSdk, _ *readline.Instance, args []string) error {
	if len(args) < 2 {
		fmt.Println("Usage: contacts add <name> <payment_identifier>")
		return nil
	}

	result, err := sdk.AddContact(breez_sdk_spark.AddContactRequest{
		Name:              args[0],
		PaymentIdentifier: args[1],
	})
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- update ---

func handleContactUpdate(sdk *breez_sdk_spark.BreezSdk, _ *readline.Instance, args []string) error {
	if len(args) < 3 {
		fmt.Println("Usage: contacts update <id> <name> <payment_identifier>")
		return nil
	}

	result, err := sdk.UpdateContact(breez_sdk_spark.UpdateContactRequest{
		Id:                args[0],
		Name:              args[1],
		PaymentIdentifier: args[2],
	})
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- delete ---

func handleContactDelete(sdk *breez_sdk_spark.BreezSdk, _ *readline.Instance, args []string) error {
	if len(args) < 1 {
		fmt.Println("Usage: contacts delete <id>")
		return nil
	}

	if err := liftError(sdk.DeleteContact(args[0])); err != nil {
		return err
	}
	fmt.Println("Contact deleted successfully")
	return nil
}

// --- list ---

func handleContactList(sdk *breez_sdk_spark.BreezSdk, _ *readline.Instance, args []string) error {
	fs := flag.NewFlagSet("contacts list", flag.ContinueOnError)
	offset := fs.Uint("offset", 0, "Number of contacts to skip")
	limit := fs.Uint("limit", 0, "Maximum number of contacts to return")
	if err := fs.Parse(args); err != nil {
		return err
	}

	req := breez_sdk_spark.ListContactsRequest{}
	if *offset > 0 {
		o := uint32(*offset)
		req.Offset = &o
	}
	if *limit > 0 {
		l := uint32(*limit)
		req.Limit = &l
	}

	result, err := sdk.ListContacts(req)
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

File: crates/breez-sdk/bindings/examples/cli/langs/csharp/Contacts.cs
using Breez.Sdk.Spark;

namespace BreezCli;

/// <summary>
/// Represents a single contacts subcommand.
/// </summary>
public class ContactCliCommand
{
    public required string Name { get; init; }
    public required string Description { get; init; }
    public required Func<BreezSdk, Func<string, string?>, string[], Task> Run { get; init; }
}

/// <summary>
/// Contacts subcommand names (used for REPL completion).
/// </summary>
public static class ContactCommandNames
{
    public static readonly string[] All =
    {
        "contacts add",
        "contacts update",
        "contacts delete",
        "contacts list",
    };
}

/// <summary>
/// Contact subcommand handlers.
/// </summary>
public static class ContactCommands
{
    /// <summary>
    /// Builds the contacts subcommand registry.
    /// </summary>
    public static Dictionary<string, ContactCliCommand> BuildRegistry()
    {
        return new Dictionary<string, ContactCliCommand>
        {
            ["add"] = new()
            {
                Name = "add",
                Description = "Add a new contact",
                Run = HandleAdd
            },
            ["update"] = new()
            {
                Name = "update",
                Description = "Update an existing contact",
                Run = HandleUpdate
            },
            ["delete"] = new()
            {
                Name = "delete",
                Description = "Delete a contact",
                Run = HandleDelete
            },
            ["list"] = new()
            {
                Name = "list",
                Description = "List contacts",
                Run = HandleList
            },
        };
    }

    /// <summary>
    /// Dispatches a contacts subcommand.
    /// </summary>
    public static async Task DispatchCommand(
        string[] args,
        BreezSdk sdk,
        Func<string, string?> readline)
    {
        var registry = BuildRegistry();

        if (args.Length == 0 || args[0] == "help")
        {
            Console.WriteLine();
            Console.WriteLine("Contacts subcommands:");
            var names = registry.Keys.OrderBy(k => k).ToList();
            foreach (var name in names)
            {
                Console.WriteLine($"  contacts {name,-30} {registry[name].Description}");
            }
            Console.WriteLine();
            return;
        }

        var subName = args[0];
        var subArgs = args.Skip(1).ToArray();

        if (!registry.TryGetValue(subName, out var cmd))
        {
            Console.WriteLine($"Unknown contacts subcommand: {subName}. Use 'contacts help' for available commands.");
            return;
        }

        await cmd.Run(sdk, readline, subArgs);
    }

    // -----------------------------------------------------------------------
    // Contacts command handlers
    // -----------------------------------------------------------------------

    // --- add ---

    private static async Task HandleAdd(BreezSdk sdk, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Usage: contacts add <name> <payment_identifier>");
            return;
        }

        var result = await sdk.AddContact(request: new AddContactRequest(
            name: args[0],
            paymentIdentifier: args[1]
        ));
        Serialization.PrintValue(result);
    }

    // --- update ---

    private static async Task HandleUpdate(BreezSdk sdk, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 3)
        {
            Console.WriteLine("Usage: contacts update <id> <name> <payment_identifier>");
            return;
        }

        var result = await sdk.UpdateContact(request: new UpdateContactRequest(
            id: args[0],
            name: args[1],
            paymentIdentifier: args[2]
        ));
        Serialization.PrintValue(result);
    }

    // --- delete ---

    private static async Task HandleDelete(BreezSdk sdk, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: contacts delete <id>");
            return;
        }

        await sdk.DeleteContact(id: args[0]);
        Console.WriteLine("Contact deleted successfully");
    }

    // --- list ---

    private static async Task HandleList(BreezSdk sdk, Func<string, string?> readline, string[] args)
    {
        uint? offset = null;
        uint? limit = null;

        if (args.Length >= 1)
        {
            offset = uint.Parse(args[0]);
        }
        if (args.Length >= 2)
        {
            limit = uint.Parse(args[1]);
        }

        var result = await sdk.ListContacts(request: new ListContactsRequest(
            offset: offset,
            limit: limit
        ));
        Serialization.PrintValue(result);
    }
}
